### DISCUSSION TITLE
Selecting WalletConnect in embedded DApp browser

### DISCUSSION THREAD

**Discussion by nikitaeverywhere** [WalletConnect/walletconnect-monorepo#1778] [Developer Support]
Hello community!

As you all know, almost every mobile wallet supports embedded DApp browser.

**Problem**: if the user selects WalletConnect as an option [there](https://react-app.walletconnect.com/), they are presented with a list of wallets which is odd. Moreover, some embedded browser's functionality like "selected network" becomes N/A, since now the communication happens via WalletConnect.

(for example, visiting [1inch](https://app.1inch.io/) and [v2 demo DApp](https://react-app.walletconnect.com/) from [Trust Wallet](https://trustwallet.com/))
<img src="https://user-images.githubusercontent.com/4989256/208099513-7fd6b323-9cbe-4621-bb1b-97745cc753a1.png" width="300"/>
Then you get a select:
<img src="https://user-images.githubusercontent.com/4989256/208100112-af198f38-7c93-482d-999d-d87ea773009c.png" width="300"/>

**Question**: are there any known (incl. JavaScript) workarounds to solve the problem?

If it turns out there's none, this issue might need to be converted to the WalletConnect API extension proposal or EIP proposal to surface:

- some sort of a JS hint for embedded WalletConnect that would allow to skip wallet selection and deeplink/connect straight away.
- some sort of a JS prop/event for embedded browser to understand it's now on WalletConnect and should not render various non-WalletConnect features like the network switch.

I can draft more ideas on how this can be done. At now I'm just looking for the community feedback.

Thanks!

**Comment by xzilja** [comment]
Hey @ZitRos 

We should be handling this flow already if wallet browser injects itself as an extension. I think Trust Wallet might not be doing this or at least is not doing this in the same way as they do with their browser extension.

Below is example of wallet inside MetaMasks iOS app working with this flow.

https://user-images.githubusercontent.com/3154053/208107032-418272df-fc63-4f89-9450-b5f7020a67f1.MP4

**Comment by nikitaeverywhere** [comment]
Thanks @0xAsimetriq for the response! This is exactly the same thing how it's handled in Trust Wallet and other embedded DApp browsers in Wallets.

I think I need to highlight these in the problem statement:

> **Problem**: if the user selects WalletConnect as an option [there](https://react-app.walletconnect.com/), **_they are presented with a list of wallets which is odd_**. Moreover, **_some embedded browser's functionality like "selected network" becomes N/A_**, since now the communication happens via WalletConnect.

(...and it will be great from the DApp browser's perspective to know that somehow)

Observe this thread as the UX improvement which will benefit both WalletConnect and DApp browser if solved:

- WC could just skip showing wallet selection (and overall, some wallets could connect via WC automatically without any prompts, like they do with `window.ethereum.request()` - Coinbase Wallet for instance)
- DApp browser could hide networks select or draw it properly, based on WC session

**Reply by xzilja** [reply]
Actually we already handle Trust as well. I think I know what's happening here - if you test on our react-dapp, it uses `@web3modal/standalone` that does not include handling for injected browsers / other connectors like coinbase etc...

Dapps that use this package are expected to manage these themselves. Thus you see that separate TrustWallet icon before WalletConnect. In standalone mode there is also no account modal view etc..., it just provides basic ui for walletconnect connecion.

Our other fully managed packages like `@web3modal/react` and `@web3modal/html` will handle all of this out of the box.

The only issue I noticed is that I need to de-duplicate Trust's injected connector from one we fetch in our explorer api (thats why there are 2)

https://user-images.githubusercontent.com/3154053/208112410-b4bb32ca-18db-4b41-a2d6-538ed063be23.MP4

**Comment by nikitaeverywhere** [comment]
Got it. Let me ask this in a different way.

1. Why would you even show a list of other wallets to choose from, if you can detect which one is installed and connect to it right away?

This is the UX question. Inside the embedded DApp browser, I believe 0% of users would be choosing another wallet app.
